File: leetcode/singleNumber/singleNumber_v2.py
class Solution:
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        res = nums[0]
        for num in nums[1:]:
            res = res ^ num
        return res

if __name__ == '__main__':
    solver = Solution()
    print(solver.singleNumber([2,2,1]))
    print(solver.singleNumber([4,1,2,1,2]))

File: leetcode/singleNumber/singleNumber_v2.cpp
#include <vector>
using namespace std;
// XOR
// * 1^0 = 1 
// * 0^1 = 1 
// * n^n = 0 (一个数与自己亦或等于0) 
// * 0^n = n (0与其他数亦或等于那个数）
class Solution {
public:
    int singleNumber(vector<int>& nums) {
        int len = nums.size();
        int res = nums[0];
        for(int i=1; i<len; ++i){
            res = res ^ nums[i];
        }
        return res;
    }
};
